File: server/storage.ts
import { type User, type InsertUser, type MathProblem, type InsertMathProblem, type FeedbackEntry, type InsertFeedback, type SystemActivity, type InsertActivity } from "@shared/schema";
import { randomUUID } from "crypto";

export interface IStorage {
  // User operations
  getUser(id: string): Promise<User | undefined>;
  getUserByUsername(username: string): Promise<User | undefined>;
  createUser(user: InsertUser): Promise<User>;
  
  // Math problem operations
  getMathProblem(id: string): Promise<MathProblem | undefined>;
  getMathProblemsByCategory(category: string): Promise<MathProblem[]>;
  searchMathProblems(query: string): Promise<MathProblem[]>;
  createMathProblem(problem: InsertMathProblem): Promise<MathProblem>;
  
  // Feedback operations
  createFeedback(feedback: InsertFeedback): Promise<FeedbackEntry>;
  getFeedbackByProblemId(problemId: string): Promise<FeedbackEntry[]>;
  
  // Activity operations
  createActivity(activity: InsertActivity): Promise<SystemActivity>;
  getRecentActivity(limit?: number): Promise<SystemActivity[]>;
  
  // Statistics
  getKnowledgeBaseStats(): Promise<{
    total: number;
    calculus: number;
    algebra: number;
    geometry: number;
    statistics: number;
  }>;
}

export class MemStorage implements IStorage {
  private users: Map<string, User>;
  private mathProblems: Map<string, MathProblem>;
  private feedbackEntries: Map<string, FeedbackEntry>;
  private systemActivities: Map<string, SystemActivity>;

  constructor() {
    this.users = new Map();
    this.mathProblems = new Map();
    this.feedbackEntries = new Map();
    this.systemActivities = new Map();
    
    // Initialize with sample math problems
    this.initializeSampleProblems();
  }

  private initializeSampleProblems() {
    const sampleProblems = [
      {
        id: randomUUID(),
        problem: "Solve the quadratic equation x² + 5x + 6 = 0",
        solution: {
          steps: [
            {
              step: 1,
              title: "Factor the quadratic expression",
              content: "x² + 5x + 6 = (x + 2)(x + 3)",
              explanation: "Find two numbers that multiply to 6 and add to 5: 2 and 3"
            },
            {
              step: 2,
              title: "Set each factor equal to zero",
              content: "x + 2 = 0  or  x + 3 = 0",
              explanation: "Use the zero product property"
            },
            {
              step: 3,
              title: "Solve for x",
              content: "x = -2  or  x = -3",
              explanation: "These are the roots of the quadratic equation"
            }
          ],
          finalAnswer: "x = -2 or x = -3"
        },
        category: "algebra",
        difficulty: "intermediate",
        source: "knowledge_base",
        createdAt: new Date()
      },
      {
        id: randomUUID(),
        problem: "Find the derivative of f(x) = 3x³ + 2x² - 5x + 1",
        solution: {
          steps: [
            {
              step: 1,
              title: "Identify the function and apply the power rule",
              content: "f(x) = 3x³ + 2x² - 5x + 1",
              explanation: "We'll differentiate each term separately using the power rule: d/dx[xⁿ] = n·xⁿ⁻¹"
            },
            {
              step: 2,
              title: "Differentiate each term",
              content: "d/dx[3x³] = 9x²\nd/dx[2x²] = 4x\nd/dx[-5x] = -5\nd/dx[1] = 0",
              explanation: "Apply the power rule to each term"
            },
            {
              step: 3,
              title: "Combine the results",
              content: "f'(x) = 9x² + 4x - 5",
              explanation: "Sum all the derivatives"
            }
          ],
          finalAnswer: "f'(x) = 9x² + 4x - 5"
        },
        category: "calculus",
        difficulty: "intermediate",
        source: "knowledge_base",
        createdAt: new Date()
      },
      {
        id: randomUUID(),
        problem: "Calculate the area of a triangle with sides 3, 4, and 5",
        solution: {
          steps: [
            {
              step: 1,
              title: "Identify triangle type",
              content: "3² + 4² = 9 + 16 = 25 = 5²",
              explanation: "This is a right triangle (Pythagorean theorem satisfied)"
            },
            {
              step: 2,
              title: "Apply area formula",
              content: "Area = ½ × base × height = ½ × 3 × 4 = 6",
              explanation: "For a right triangle, use the two perpendicular sides as base and height"
            }
          ],
          finalAnswer: "Area = 6 square units"
        },
        category: "geometry",
        difficulty: "basic",
        source: "knowledge_base",
        createdAt: new Date()
      }
    ];

    sampleProblems.forEach(problem => {
      this.mathProblems.set(problem.id, problem as MathProblem);
    });
  }

  async getUser(id: string): Promise<User | undefined> {
    return this.users.get(id);
  }

  async getUserByUsername(username: string): Promise<User | undefined> {
    return Array.from(this.users.values()).find(
      (user) => user.username === username,
    );
  }

  async createUser(insertUser: InsertUser): Promise<User> {
    const id = randomUUID();
    const user: User = { ...insertUser, id };
    this.users.set(id, user);
    return user;
  }

  async getMathProblem(id: string): Promise<MathProblem | undefined> {
    return this.mathProblems.get(id);
  }

  async getMathProblemsByCategory(category: string): Promise<MathProblem[]> {
    return Array.from(this.mathProblems.values()).filter(
      problem => problem.category === category
    );
  }

  async searchMathProblems(query: string): Promise<MathProblem[]> {
    const lowercaseQuery = query.toLowerCase();
    return Array.from(this.mathProblems.values()).filter(
      problem => problem.problem.toLowerCase().includes(lowercaseQuery)
    );
  }

  async createMathProblem(insertProblem: InsertMathProblem): Promise<MathProblem> {
    const id = randomUUID();
    const problem: MathProblem = { 
      ...insertProblem, 
      id,
      createdAt: new Date()
    };
    this.mathProblems.set(id, problem);
    return problem;
  }

  async createFeedback(insertFeedback: InsertFeedback): Promise<FeedbackEntry> {
    const id = randomUUID();
    const feedback: FeedbackEntry = { 
      ...insertFeedback, 
      id,
      createdAt: new Date(),
      problemId: insertFeedback.problemId || null,
      comments: insertFeedback.comments || null
    };
    this.feedbackEntries.set(id, feedback);
    return feedback;
  }

  async getFeedbackByProblemId(problemId: string): Promise<FeedbackEntry[]> {
    return Array.from(this.feedbackEntries.values()).filter(
      feedback => feedback.problemId === problemId
    );
  }

  async createActivity(insertActivity: InsertActivity): Promise<SystemActivity> {
    const id = randomUUID();
    const activity: SystemActivity = { 
      ...insertActivity, 
      id,
      createdAt: new Date(),
      details: insertActivity.details || null
    };
    this.systemActivities.set(id, activity);
    return activity;
  }

  async getRecentActivity(limit: number = 10): Promise<SystemActivity[]> {
    const activities = Array.from(this.systemActivities.values())
      .sort((a, b) => b.createdAt!.getTime() - a.createdAt!.getTime());
    return activities.slice(0, limit);
  }

  async getKnowledgeBaseStats(): Promise<{
    total: number;
    calculus: number;
    algebra: number;
    geometry: number;
    statistics: number;
  }> {
    const problems = Array.from(this.mathProblems.values());
    return {
      total: problems.length,
      calculus: problems.filter(p => p.category === 'calculus').length,
      algebra: problems.filter(p => p.category === 'algebra').length,
      geometry: problems.filter(p => p.category === 'geometry').length,
      statistics: problems.filter(p => p.category === 'statistics').length,
    };
  }
}

export const storage = new MemStorage();

File: backend/services/storage.py
"""
Enhanced Storage Service for the Math Routing Agent.
Handles persistent storage of problems, feedback, and system activities.
"""

import json
import asyncio
from datetime import datetime
from typing import List, Dict, Any, Optional
from uuid import uuid4

class StorageService:
    def __init__(self):
        # In-memory storage (replace with actual database in production)
        self.math_problems = {}
        self.feedback_entries = {}
        self.system_activities = {}
        self.users = {}
        
    # Math Problems
    async def create_math_problem(
        self,
        problem: str,
        solution: Dict[str, Any],
        category: str,
        difficulty: str,
        source: str
    ) -> Dict[str, Any]:
        """Create a new math problem entry."""
        problem_id = str(uuid4())
        problem_entry = {
            "id": problem_id,
            "problem": problem,
            "solution": solution,
            "category": category,
            "difficulty": difficulty,
            "source": source,
            "created_at": datetime.now()
        }
        
        self.math_problems[problem_id] = problem_entry
        return problem_entry
    
    async def get_math_problem(self, problem_id: str) -> Optional[Dict[str, Any]]:
        """Get a math problem by ID."""
        return self.math_problems.get(problem_id)
    
    async def get_math_problems_by_category(self, category: str) -> List[Dict[str, Any]]:
        """Get all math problems in a specific category."""
        return [
            problem for problem in self.math_problems.values()
            if problem["category"] == category
        ]
    
    async def search_math_problems(self, query: str) -> List[Dict[str, Any]]:
        """Search math problems by text content."""
        query_lower = query.lower()
        return [
            problem for problem in self.math_problems.values()
            if query_lower in problem["problem"].lower()
        ]
    
    # Feedback Entries
    async def create_feedback(
        self,
        problem_id: str,
        accuracy_rating: int,
        clarity_rating: str,
        comments: Optional[str] = None,
        is_helpful: bool = True
    ) -> Dict[str, Any]:
        """Create a new feedback entry."""
        feedback_id = str(uuid4())
        feedback_entry = {
            "id": feedback_id,
            "problem_id": problem_id,
            "accuracy_rating": accuracy_rating,
            "clarity_rating": clarity_rating,
            "comments": comments,
            "is_helpful": is_helpful,
            "created_at": datetime.now()
        }
        
        self.feedback_entries[feedback_id] = feedback_entry
        return feedback_entry
    
    async def get_feedback_by_problem_id(self, problem_id: str) -> List[Dict[str, Any]]:
        """Get all feedback for a specific problem."""
        return [
            feedback for feedback in self.feedback_entries.values()
            if feedback["problem_id"] == problem_id
        ]
    
    async def get_feedback_stats(self) -> Dict[str, Any]:
        """Get feedback statistics."""
        if not self.feedback_entries:
            return {"total": 0, "average_rating": 0, "helpful_percentage": 0}
        
        total = len(self.feedback_entries)
        total_rating = sum(f["accuracy_rating"] for f in self.feedback_entries.values())
        helpful_count = sum(1 for f in self.feedback_entries.values() if f["is_helpful"])
        
        return {
            "total": total,
            "average_rating": total_rating / total if total > 0 else 0,
            "helpful_percentage": (helpful_count / total * 100) if total > 0 else 0
        }
    
    # System Activities
    async def create_activity(
        self,
        action: str,
        source: str,
        details: Optional[str] = None
    ) -> Dict[str, Any]:
        """Create a new system activity entry."""
        activity_id = str(uuid4())
        activity_entry = {
            "id": activity_id,
            "action": action,
            "source": source,
            "details": details,
            "created_at": datetime.now()
        }
        
        self.system_activities[activity_id] = activity_entry
        return activity_entry
    
    async def get_recent_activity(self, limit: int = 10) -> List[Dict[str, Any]]:
        """Get recent system activities."""
        activities = list(self.system_activities.values())
        activities.sort(key=lambda x: x["created_at"], reverse=True)
        return activities[:limit]
    
    async def get_activity_by_source(self, source: str) -> List[Dict[str, Any]]:
        """Get activities filtered by source."""
        return [
            activity for activity in self.system_activities.values()
            if activity["source"] == source
        ]
    
    # Knowledge Base Statistics
    async def get_knowledge_base_stats(self) -> Dict[str, int]:
        """Get comprehensive knowledge base statistics."""
        if not self.math_problems:
            return {
                "total": 0,
                "calculus": 0,
                "algebra": 0,
                "geometry": 0,
                "statistics": 0,
                "trigonometry": 0
            }
        
        problems = list(self.math_problems.values())
        stats = {
            "total": len(problems),
            "calculus": len([p for p in problems if p["category"] == "calculus"]),
            "algebra": len([p for p in problems if p["category"] == "algebra"]),
            "geometry": len([p for p in problems if p["category"] == "geometry"]),
            "statistics": len([p for p in problems if p["category"] == "statistics"]),
            "trigonometry": len([p for p in problems if p["category"] == "trigonometry"])
        }
        
        return stats
    
    # User Management (basic implementation)
    async def create_user(self, username: str, password: str) -> Dict[str, Any]:
        """Create a new user."""
        user_id = str(uuid4())
        user_entry = {
            "id": user_id,
            "username": username,
            "password": password,  # In production, hash this!
            "created_at": datetime.now()
        }
        
        self.users[user_id] = user_entry
        return user_entry
    
    async def get_user_by_username(self, username: str) -> Optional[Dict[str, Any]]:
        """Get user by username."""
        return next(
            (user for user in self.users.values() if user["username"] == username),
            None
        )
    
    # Data Export/Import for backup
    async def export_data(self) -> Dict[str, Any]:
        """Export all data for backup."""
        return {
            "math_problems": {
                k: {**v, "created_at": v["created_at"].isoformat()} 
                for k, v in self.math_problems.items()
            },
            "feedback_entries": {
                k: {**v, "created_at": v["created_at"].isoformat()}
                for k, v in self.feedback_entries.items()
            },
            "system_activities": {
                k: {**v, "created_at": v["created_at"].isoformat()}
                for k, v in self.system_activities.items()
            },
            "users": {
                k: {**v, "created_at": v["created_at"].isoformat()}
                for k, v in self.users.items()
            }
        }
    
    async def import_data(self, data: Dict[str, Any]):
        """Import data from backup."""
        for category, items in data.items():
            if category == "math_problems":
                self.math_problems = {
                    k: {**v, "created_at": datetime.fromisoformat(v["created_at"])}
                    for k, v in items.items()
                }
            elif category == "feedback_entries":
                self.feedback_entries = {
                    k: {**v, "created_at": datetime.fromisoformat(v["created_at"])}
                    for k, v in items.items()
                }
            elif category == "system_activities":
                self.system_activities = {
                    k: {**v, "created_at": datetime.fromisoformat(v["created_at"])}
                    for k, v in items.items()
                }
            elif category == "users":
                self.users = {
                    k: {**v, "created_at": datetime.fromisoformat(v["created_at"])}
                    for k, v in items.items()
                }